### PR TITLE
Reworked aggregators.

### PR DESCRIPTION
**Pull request by tpoterba**
Many changes to aggregables
 - Aggregables have an element type
 - Added .map and .filter opterations on aggregables
 - Most aggregator functions do not take arguments
 - Added 'sum()' aggregator to sum numeric or Array[Numeric]

 Added oneHotGt function to genotype

### CODE REVIEW

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 1]
```diff
@@ -1017,35 +1020,75 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
...
           }
           case error =>
-            parseError(s"method `$method' expects a lambda function [param => Any], got invalid mapping [param => $error]")
+            parseError(s"method `$method' expects a lambda function (param => Any), got invalid mapping (param => $error)")
+        }
+
+      case (agg: TAggregable, "map", rhs) =>
+        val (param, body) = rhs match {
+          case Array(Lambda(_, p, b)) => (p, b)
+        }
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: You can just write `val Array(Lambda(_, param, body)) = rhs`.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 2]
```diff
@@ -1017,35 +1020,75 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
...
+        val localA = agg.ec.a
+        localA += null
+        val st = agg.ec.st + ((param, (localIdx, agg.elementType)))
+        body.typecheck(agg.ec.copy(st = st))
+        `type` = body.`type` match {
+          case TBoolean =>
+            val fn = body.eval(agg.ec.copy(st = st))
+            val filterF = (a: Any) => {
+              localA(localIdx) = a
+              //              println(s"in filterF, a=$a")
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: Remove.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 3]
```diff
@@ -1357,185 +1362,220 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
...
+      val seqOp = (a: Any, next: Any) => {
+        val (value, included) = aggF(next)
+        if (included) {
+          val sc = a.asInstanceOf[StatCounter]
+          if (value != null)
+            t match {
+              case TInt => sc.merge(value.asInstanceOf[Int])
+              case TLong => sc.merge(value.asInstanceOf[Long])
+              case TFloat => sc.merge(value.asInstanceOf[Float])
+              case TDouble => sc.merge(value.asInstanceOf[Double])
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: `StatCounter.merge` just takes a Double, so you might as well just call the numeric conversion code to Double here.

File: src/main/scala/org/broadinstitute/hail/variant/Genotype.scala [hunk 4]
```diff
@@ -173,6 +173,16 @@ class Genotype(private val _gt: Int,
 
   def nNonRefAlleles: Option[Int] = Genotype.nNonRefAlleles(_gt)
 
+  def oneHotGt(v: Variant): Option[Array[Int]] = {
+    gt.map { call =>
+      val gtPair = Genotype.gtPair(call)
+      val a = Array.fill(v.nAlleles)(0)
+      a(gtPair.j) += 1
+      a(gtPair.k) += 1
+      a
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/variant/Genotype.scala]: `oneHotGt` sums to two?  I was expecting a vector over genotypes.  That's confusing.  How about `oneHotAlelles`?

File: src/main/scala/org/broadinstitute/hail/variant/Genotype.scala [hunk 5]
```diff
@@ -194,15 +204,18 @@ class Genotype(private val _gt: Int,
...
   }
 
-  def pAB(theta: Double = 0.5): Option[Double] = ad.map { case Array(refDepth, altDepth) =>
-    val d = new BinomialDistribution(refDepth + altDepth, theta)
-    val minDepth = refDepth.min(altDepth)
-    val minp = d.probability(minDepth)
-    val mincp = d.cumulativeProbability(minDepth)
-    (2 * mincp - minp).min(1.0).max(0.0)
+  def pAB(theta: Double = 0.5): Option[Double] = ad.map {
+    case Array(refDepth, altDepth) =>
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/variant/Genotype.scala]: Any reason in particular you changed the indentation here?  For anonymous pattern matching functions with one alternative, I prefer the case just after the open curly brace.

File: src/main/scala/org/broadinstitute/hail/expr/package.scala [hunk 6]
```diff
@@ -3,7 +3,7 @@ package org.broadinstitute.hail
 package object expr {
   type SymbolTable = Map[String, (Int, BaseType)]
 
-  type Aggregator = (() => Any, (Any) => Any, (Any, Any) => Any, Int)
+  type Aggregator = (() => Any, (Any, Any) => Any, (Any, Any) => Any, Int)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/package.scala]: I think this should be a typed case class.  How about:

```
type Aggregator = TypedAggregator[Any, Any]
abstract class TypedAggregator[T, S] { self =>
  def zero: S
  def seqOp(x: T, acc: S): S
  def combOp...
  def erase: Aggregator = new Aggregator[Any, Any] {
    def zero = self.zero
    def seqOp(x: Any, acc: Any) = self.seqOp(x.asInstanceOf[T], acc.asInstanceOf[S])
    ...
  }
}
```

Then below when you create aggregators, you can just say `new TypedAggregator[T, S] { def zero = ... }`.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 7]
```diff
@@ -1357,185 +1362,220 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
...
+            }
           sc
+        } else a
       }
 
       val combOp = (a: Any, b: Any) => a.asInstanceOf[StatCounter].merge(b.asInstanceOf[StatCounter])
 
-      val recast: (Double) => Any = localType match {
+      val recast: (Double) => Any = t match {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: I never liked this recast stuff.  I think `.stats` should return all doubles or we should add our own typed StatCounter (of type TypedAggregator!)

File: src/main/scala/org/broadinstitute/hail/expr/Type.scala [hunk 8]
```diff
@@ -230,13 +231,47 @@ case object TString extends Type {
...
+  override def toString: String = s"Aggregable[${ elementType.toString }]"
 
-case class TAggregable(ec: EvalContext) extends BaseType {
-  override def toString = "Aggregable"
+  def ec: EvalContext
 
-  def typeCheck(a: Any): Boolean = a == null || a.isInstanceOf[Iterable[_]]
+  def elementType: Type
+
+  def f: (Any) => (Any, Boolean)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/Type.scala]: Any not `=> Option[Any]`?

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 9]
```diff
@@ -1357,185 +1362,220 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
...
+
+    case (agg: TAggregable, "sum", Array()) =>
+      val localIdx = agg.ec.a.length
+      val localA = agg.ec.a
+      localA += null
+
+      val aggF = agg.f
+
+      val (zv, seqOp, combOp) = (`type`: @unchecked) match {
+        case TLong =>
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: This should just be something like `new SumAggregator[Long]`.

File: src/main/scala/org/broadinstitute/hail/expr/package.scala (hunk 6 above)
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/package.scala]: You might switch the names and make the typed version `Aggregator`.  I thought this a very reasonable object and, for example, the VSM aggregation functions should take an aggregator.  (But don't add that to this PR.)

File: src/main/scala/org/broadinstitute/hail/expr/Type.scala [hunk 10]
```diff
@@ -230,13 +231,47 @@ case object TString extends Type {
...
+}
+
+case class BaseAggregable(ec: EvalContext, elementType: Type) extends TAggregable {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/Type.scala]: I'd also type the aggregators up.  Seems like it would be cleaner.

**Comment by cseed** [comment]
Awesome changed, great functionality.  Back to you.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala (hunk 1 above)
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: I want to check that it's a lambda function, though.  I think the code you suggested will produce bad error messages

File: src/main/scala/org/broadinstitute/hail/variant/Genotype.scala (hunk 5 above)
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/variant/Genotype.scala]: I think this was autoformat.  Changed it back.

**Comment by tpoterba** [comment]
Addressed comments.  I wasn't able to easily type the aggregables, though, since they still go through arbitrary exprs.

**Comment by tpoterba** [comment]
However, I think things are looking a lot nicer!

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 11]
```diff
@@ -1265,6 +1258,10 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
           }
       }
 
+    case (agg: TAggregable, "map", Array(Lambda(_, param, body))) => () => null
+
+    case (agg: TAggregable, "filter", Array(Lambda(_, param, body))) => () => null
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: I don't quite get these.  Shouldn't it be an error to evaluate map and filter on aggregables?

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: Good point.  It is caught at type check so we could probably delete these

**Comment by cseed** [comment]
Otherwise, looks good.  Delete those guys and I'll merge it in.

**Pull request merged**